File: 1-trebuchet-pt2.c
#include <stdio.h>
#include <stdlib.h>
#include <stdint.h>
#include <stdbool.h>
#include <string.h>

// Second half solution: 53348

#define MAX_DIGIT_LENGTH 2     // maximum digit size (always 2 in this case)
#define MAX_BUFFER_SIZE   255  // maximum line length
#define NUMBER_WORD_COUNT 9    // amount of number words

// constant for each wrod number
char* NUMBER_WORDS[NUMBER_WORD_COUNT] = {
    "one", "two", "three", "four", "five",
    "six", "seven", "eight", "nine"
};

// function for updating first and second digit
void set_digits(char* digit1, char* digit2, char digit) {

    // set the first digit if not set
    if (*digit1 == 0)
        *digit1 = digit;

    // always set the second digit
    *digit2 = digit;

}

// function for checking if texts starts with a word
bool starts_with(char* line, char* word, uint32_t index) {

    // while the pointer is still pointing to the word's content
    while (*word) {

        // check if the current word's character matches the text's
        // current character
        if (*word != line[index])
            return false;

        // increment word pointer and index
        word++;
        index++;
    }

    // starts with word
    return true;
}

// function for decoding a line
uint8_t decode_line(char* line) {

    // first and second digit
    char digit1 = 0, digit2 = 0;

    // loop through each char in line
    for (uint32_t i = 0; i < strlen(line); i++) {

        // get the ascii value for the char
        uint32_t ascii = line[i];

        // check if it's in range for a number
        if (ascii > 48 && ascii < 58)
            set_digits(&digit1, &digit2, line[i]);
        else { // otherwise try to decode word number

            // loop through each word in the list
            for (uint8_t wordIndex = 0; wordIndex < NUMBER_WORD_COUNT; wordIndex++) {

                // set digits if the line starts with the word
                bool result = starts_with(line, NUMBER_WORDS[wordIndex], i);
                if (result)
                    set_digits(&digit1, &digit2, (wordIndex + 1) + '0');

            }
            
        }
    
    }

    // concat both digits together by serializing them into the string
    char decoded[MAX_DIGIT_LENGTH];
    sprintf(decoded, "%c%c", digit1, digit2);

    // return back into a number
    return atoi(decoded);
}

int main(int argc, char** argv) {

    // open input file from command line argument
    FILE* infile = fopen(argv[1], "r");

    // define total and line buffer
    uint32_t total = 0;
    char line[MAX_BUFFER_SIZE];

    // loop through each line in file
    while (fgets(line, MAX_BUFFER_SIZE, infile)) {

        // decode line and add to total
        uint8_t count = decode_line(line);
        total += count;

    }

    // print total
    printf("Total: %d\n", total);

    // close input file and return success
    fclose(infile);
    return 0;
}

File: 1-trebuchet-pt2.py
import sys  # used to gather command line arguments

# Second half solution: 53348

# define a constant of number words
# their index in the list will be their value
NUMBER_WORDS = [
    'one', 'two', 'three', 'four', 'five',
    'six', 'seven', 'eight', 'nine'
]

def decode_line(line: str) -> int:

    # list to store digits
    digits = []

    # character index
    i = 0

    # loop through each character in line
    for c in line:

        # decode if digit
        if c.isdigit():
            digits.append(c)

        # attempt to decode number word
        else:

            # iterate through each number word
            wordIndex = 1
            for word in NUMBER_WORDS:

                # append number word as digit if current line index
                # starts with number word
                if line[i:].startswith(word):
                    digits.append(wordIndex)
                wordIndex += 1

        # increment character index
        i += 1

    # serialize the first and last digit together
    serialized = f'{digits[0]}{digits[len(digits) - 1]}'

    # return as a converted number
    return int(serialized)

# main entrypoint
def main():

    # open input file from command line argument
    infile = open(sys.argv[1], 'r')

    # read and decode each line, add to total
    total = 0
    for line in infile.readlines():
        decoded = decode_line(line)
        total += decoded

    # print total and close input file
    print(f'Total: {total}')
    infile.close()

# run if python script is directly executed
if __name__ == '__main__':
    main()